src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
```

Comment:
Have you examined named locks @sfmskywalker ? Because as we discussed having a single lock object like this will queue up all the acquire calls, regardless of context, and now e.g. a background task executing will wait until a foreground content item publish finishes when both use locking through IDistributedLockService.